app/services/user_locale_setter.rb

```diff
@@ -0,0 +1,62 @@
+# frozen_string_literal: true
+
+class UserLocaleSetter
+  def initialize(current_user, params_locale, cookies)
+    @current_user = current_user
+    @params_locale = params_locale
+    @cookies = cookies
+  end
+
+  def set_locale
+    save_locale_from_params
+
+    I18n.locale = valid_current_locale
+  end
+
+  def ensure_valid_locale_persisted
+    return unless current_user && !available_locale?(current_user.locale)
+
+    current_user.update!(locale: valid_current_locale)
+  end
+
+  def self.valid_locale_for_user(user)
+    if user.andand.locale.present? && available_locale?(user.locale)
+      user.locale
+    else
+      I18n.default_locale
+    end
+  end
+
+  def self.available_locale?(locale)
```

Comment:
why class method? why not just the instance and private method?

Reply:
I had to make this available as an instance method because #valid_locale_for_user was an instance method. Updated :+1: